File: src/main/java/com/hpe/ossm/jcluster/LocalListener.java
package com.hpe.ossm.jcluster;

import akka.actor.AbstractActorWithTimers;
import akka.actor.Address;
import akka.cluster.Cluster;
import akka.cluster.ClusterEvent;
import akka.cluster.Member;
import java.time.Duration;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import com.hpe.ossm.jcluster.messages.Ping;
import com.hpe.ossm.jcluster.messages.Pong;
import com.hpe.ossm.jcluster.messages.UpdateClusterStatus;
import scala.collection.immutable.IndexedSeq;

import static akka.pattern.Patterns.ask;

public class LocalListener extends AbstractActorWithTimers {
    private Cluster cluster = Cluster.get(context().system());
    private int connectedSeedNodes = -1;

    @Override
    public void preStart() {
        cluster.registerOnMemberUp(() ->
                cluster.subscribe(self(), ClusterEvent.MemberUp.class, ClusterEvent.UnreachableMember.class, ClusterEvent.MemberRemoved.class)
        );
        timers().startPeriodicTimer("updateStatus", new UpdateClusterStatus(), Duration.ofSeconds(60));
    }

    @Override
    public void postStop() {
        cluster.unsubscribe(self());
    }

    private void reduceConnection(){
        connectedSeedNodes--;
        if (connectedSeedNodes == 0) {
            System.out.println("All Seed Nodes are down, showdown the node.");
            cluster.down(cluster.selfAddress());
        }
    }
    private void checkConnectionWithMaster(Address address) {
        CompletableFuture<Object> future = ask(
                context().actorSelection(address + "/user/localListener"),
                new Ping(), Duration.ofSeconds(5)).toCompletableFuture();
        try {
            if (!(future.join() instanceof Pong)) {
                System.out.println("Master " + address + " id down.");
                reduceConnection();
            }
        } catch (CompletionException e) {
            System.out.println("Master " + address + " id down.");
            reduceConnection();
        }
    }

    private void checkConnections(ClusterEvent.CurrentClusterState status) {
        IndexedSeq<Address> seedNodesAddress = cluster.settings().SeedNodes();

        Iterable<Member> members = status.getMembers();
        java.util.List<Address> memberAddress = new java.util.ArrayList<>();
        members.forEach(member -> memberAddress.add(member.address()));
        seedNodesAddress = seedNodesAddress.filter(memberAddress::contains).toIndexedSeq();

        connectedSeedNodes = seedNodesAddress.length();

        Set<Member> unReachNodes = status.getUnreachable();

        java.util.List<Address> unReachNodesAddress = new java.util.ArrayList<>();
        unReachNodes.forEach(node -> unReachNodesAddress.add(node.address()));


        IndexedSeq<Address> unReachSeedNodesAddress = seedNodesAddress.filter(unReachNodesAddress::contains).toIndexedSeq();

        if (unReachSeedNodesAddress.length() == seedNodesAddress.length()) {  //all seed node are unreached
            for (int i = 0; i < unReachSeedNodesAddress.length(); i++) {
                Address address = unReachSeedNodesAddress.apply(i);
                checkConnectionWithMaster(address);
            }
//or
//            unReachSeedNodesAddress.foreach(n->{
//                checkConnectionWithMaster(n);
//                return 1;
//            });
        }
    }

    public Receive createReceive() {
        return receiveBuilder()
                .match(ClusterEvent.CurrentClusterState.class, this::checkConnections)
                .match(ClusterEvent.MemberUp.class, mUp ->
                    System.out.println("Member is Up " + mUp.member().address())
                )
                .match(ClusterEvent.UnreachableMember.class, mUnreachable ->
                    System.out.println("Member is Unreachable " + mUnreachable.member().address())
                )
                .match(ClusterEvent.MemberRemoved.class, mRemoved ->
                    System.out.println("Member is Removed " + mRemoved.member())
                )
                .match(Ping.class, p -> {
                    System.out.println("Receive Ping from " + sender().path());
                    sender().tell(new Pong(), self());
                })
                .match(UpdateClusterStatus.class, m -> cluster.sendCurrentClusterState(self()))
                .match(ClusterEvent.MemberEvent.class, message -> System.out.println("[LL] Receive ignored message " + message.toString()))
                .build();

    }
}

File: src/main/scala/com/hpe/ossm/scluster/LocalListener.scala
package com.hpe.ossm.scluster

import akka.actor.{Actor, Address, Timers}
import akka.cluster.ClusterEvent.{CurrentClusterState, MemberRemoved, MemberUp, UnreachableMember}
import akka.cluster.Cluster
import akka.pattern.Patterns.ask
import com.hpe.ossm.jcluster.messages.{Ping, Pong, UpdateClusterStatus}
import scala.concurrent.ExecutionContext
import scala.concurrent.duration._

class LocalListener extends Actor with Timers {
    private val cluster = Cluster(context.system)
    private var connectedSeedNodes = -1

    implicit val ec: ExecutionContext = context.dispatcher

    override def preStart(): Unit = {
        cluster.registerOnMemberUp(() ->
            cluster.subscribe(self, classOf[MemberUp], classOf[UnreachableMember], classOf[MemberRemoved])
        )
        timers.startPeriodicTimer(
            "updateStatus", new UpdateClusterStatus, 60.seconds)
    }

    override def postStop(): Unit = cluster.unsubscribe(self)

    private def reduceConnection(): Unit = {
        connectedSeedNodes = connectedSeedNodes - 1
        if (connectedSeedNodes <= 0) {
            println("All Seed Nodes are down, shutdown the node.")
            cluster.down(cluster.selfAddress)
        }
    }

    private def checkConnectionWithMaster(address: Address): Unit = {
        try {
            ask(context.actorSelection(address + "/user/localListener"), new Ping(), 5.seconds).mapTo[Pong]
                .onComplete(r =>
                    if (r.isFailure) {
                        println(s"Seed Node $address is down.")
                        reduceConnection()
                    }
                )
        } catch {
            case _: Exception =>
                println(s"Seed Node $address is down.")
                reduceConnection()
        }
    }

    private def checkConnections(status: CurrentClusterState): Unit = {
        val seedNodeAddress = cluster.settings.SeedNodes.filter(status.members.map(_.address).contains)
        val unReachSeedNodesAddress = seedNodeAddress.filter(status.unreachable.map(_.address).contains)
        if (seedNodeAddress.length == unReachSeedNodesAddress.length) //all seed node are unreached
            seedNodeAddress.foreach(checkConnectionWithMaster)
    }

    override def receive: Receive = {
        case s: CurrentClusterState => checkConnections(s)
        case m: MemberUp => println(s"Member is up  ${m.member.address}")
        case m: UnreachableMember => println(s"Member is unreachable  ${m.member.address}")
        case m: MemberRemoved => println(s"Member is removed  ${m.member.address}")
        case _: Ping => sender ! new Pong
        case _: UpdateClusterStatus => cluster.sendCurrentClusterState(self)
        case a: Any => println(s"Received ignored message ${a.toString}")
    }
}